scripts/cocci/null-after-free.cocci

```diff
@@ -0,0 +1,84 @@
+// Read a list of 'free' type function names from stdin, if stdin is a pipe.
+@initialize:python@
+@@
+import sys
+functions = set(['free'])
+if not sys.stdin.isatty():
+    for line in sys.stdin:
+        f = line.rstrip('\n')
+        functions.add(f)
+
+// Match all function names.
+@a@
+identifier F;
+@@
+    F(...)
+
+// Match all function names in the set.
+@script:python b depends on a@
+F << a.F;
+f;
+@@
+if F in functions:
+    coccinelle.f = F
+    cocci.include_match(True)
+
+// Remove casts.
+@@
+identifier b.f;
+type t;
+expression x;
+@@
+-   f((t *)(x))
++   f(x)
+
+// Add NULL assignment after free, if absent.
+@@
+identifier b.f;
+expression x;
+@@
+(
+    f(x);
+    x = NULL;
+|
+    f(x);
++   x = NULL;
+)
+
+// Remove NULL assignments to local variables.
+@@
+local idexpression x;
+@@
+-   x = NULL;
+    ... when != x
+
+// Remove NULL assignments followed later by another assignment.
+@@
+idexpression x;
+identifier y;
+@@
+(
+-   x = NULL;
+    ... when != x
+    x = ...;
+|
+-   x->y = NULL;
+    ... when != x->y
+    x->y = ...;
+)
+
+// Remove NULL assignments followed immediately by another assignment.
+@@
+expression z;
+@@
+-   z = NULL;
+    z = ...;
+
+// Remove NULL assignments to structure members before the structure is freed.
```

Comment:
This is somewhat debatable, but I don’t mind applying it.

We sometimes intentionally zero out structure fields before freeing the whole structure in (mostly vain) attempt to prevent ‘issues’ following an actual double free. For example, in themis_secure_message_signer_destroy() function seen below, if we zero out `ctx->sign_ctx` then if `ctx` itself is freed twice then at least it’s safe to call soter_sign_destroy() twice.

I know that this is a stretch and twisted logic, but that’s some attempt to rationalize these lines.

Anyhow, given an automated way to remove such assignments, I’d personally get rid of them.

Reply:
I'd NOT apply this change before reading through all code, because sometimes we intentionally zero out and nullify structure to minimize amount of time data is stored in memory.

Reply:
This hunk deals with zeroing out pointers. We do, in fact, wipe sensitive data from memory (sometimes, not always 😞), but I doubt that there is anything sensitive in numerical address values.

Reply:
Valid statement 👍

Reply:
Oh I hadn't thought of that. It's safe to remove that rule, or accept its output very carefully. I will work on another semantic patch that calls soter_wipe before free for variables of certain types like keys.